crates/ra_hir/src/resolve.rs

```diff
@@ -0,0 +1,413 @@
+//! Name resolution.
+use std::sync::Arc;
+
+use rustc_hash::FxHashMap;
+
+use test_utils::tested_by;
+use ra_db::CrateId;
+
+use crate::{
+    ModuleDef, Module,
+    db::HirDatabase,
+    name::{Name, KnownName},
+    nameres::{PerNs, ItemMap, ReachedFixedPoint},
+    generics::GenericParams,
+    expr::{scope::{ExprScopes, ScopeId}, PatId},
+    impl_block::ImplBlock,
+    path::{Path, PathKind, PathSegment},
+};
+
+#[derive(Debug, Clone, Default)]
+pub struct Resolver<'a> {
+    scopes: Vec<Scope<'a>>,
+}
+
+// TODO how to store these best
+#[derive(Debug, Clone)]
+pub(crate) struct ModuleItemMap {
+    item_map: Arc<ItemMap>,
+    module: Module,
+}
+
+/// For usage during import resolution.
+#[derive(Debug, Clone)]
+pub(crate) struct ModuleItemMapRef<'a> {
+    item_map: &'a ItemMap,
+    module: Module,
+}
+
+#[derive(Debug, Clone)]
+pub(crate) struct ExprScope {
+    expr_scopes: Arc<ExprScopes>,
+    scope_id: ScopeId,
+}
+
+#[derive(Debug, Clone)]
+pub(crate) enum Scope<'a> {
+    /// All the items and imported names of a module
+    ModuleScope(ModuleItemMap),
+    /// All the items and imported names of a module, but with a reference to
+    /// the ItemMap. This is used during import resolution.
+    ModuleScopeRef(ModuleItemMapRef<'a>),
+    /// Brings the generic parameters of an item into scope
+    GenericParams(Arc<GenericParams>),
+    /// Brings `Self` into scope
+    ImplBlockScope(ImplBlock),
+    /// Local bindings
+    ExprScope(ExprScope),
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum Resolution {
+    // FIXME make these tuple variants
+    /// An item
+    Def {
+        def: ModuleDef,
+    },
+    /// A local binding (only value namespace)
+    LocalBinding {
+        pat: PatId,
+    },
+    /// A generic parameter
+    GenericParam {
+        idx: u32,
+    },
+    SelfType(ImplBlock),
```

Comment:
Self type feels like a kind of generic param

Reply:
Yeah, I think that's how rustc handles it as well. Currently we don't handle impl block generics at all though, I think when we implement that, we can do it that way.